Source/Dafny/Resolver.cs

```diff
@@ -11492,9 +11492,13 @@ private class RBranchExpr : RBranch {
       }
 
       if (!type.IsDatatype) {
-        if (pat is IdPattern) {
-          /* =[1]= */
-          CheckLinearVarPattern(type, (IdPattern) pat);
+        if (pat is IdPattern id) {
+          if (id.Id == BuiltIns.TupleTypeCtorNamePrefix + "0") {
+            reporter.Error(MessageSource.Resolver, pat.Tok, "case argument type does not match source argument type");
```

Comment:
What if `type` really _is_ `()` (that is, the 0-tuple type)?

Reply:
Works OK. This code is guarded by if (!type.IsDatatype), so that case does not pass through what is shown in this snippet.